pkg/resource/deploy/step.go

```diff
@@ -25,9 +25,19 @@ import (
 	"github.com/pulumi/pulumi/pkg/util/contract"
 )
 
+// StepCompleteFunc is the type of functions returned from Step.Apply. These functions are to be called
+// when the engine has fully retired a step.
+type StepCompleteFunc func()
+
 // Step is a specification for a deployment operation.
 type Step interface {
-	Apply(preview bool) (resource.Status, error) // applies or previews this step.
+	// Apply applies or previews this step. It returns the status of the resource after the step application,
+	// a function to call to signal that this step has fully completed, and an error, if one occurred while applying
+	// the step.
+	//
+	// The returned StepCompleteFunc, if not null, must be called after committing the results of this step into
+	// the state of the deployment.
+	Apply(preview bool) (resource.Status, StepCompleteFunc, error) // applies or previews this step.
```

Comment:
Did you consider instead adding another method to the `Step` interface? I don't have a strong opinion here--I actually like the model of returning a callback--but I can see some advantages to the other approach as well. E.g. if most/all `Apply` implementations will return a non-`nil` callback, it might as well be part of the interface.

Reply:
Similarly if the implementations ~always return `nil` if `err` is non-nil, it may be better to encode the relevant logic in the caller (i.e. the step executor).

Reply:
I'll investigate in more detail, but I went with the callback originally because the closure I'm returning contains some step state that I'd otherwise have to persist somewhere. I'll revisit this.